sb3_contrib/qrdqn/policies.py

```diff
@@ -0,0 +1,244 @@
+        self.features_dim = features_dim
+        self.n_quantiles = n_quantiles
+        self.normalize_images = normalize_images
+        action_dim = self.action_space.n  # number of actions
+        quantile_net = create_mlp(self.features_dim, action_dim * self.n_quantiles, self.net_arch, self.activation_fn)
+        self.quantile_net = nn.Sequential(*quantile_net)
+
+    def forward(self, obs: th.Tensor) -> th.Tensor:
+        """
+        Predict the quantile-values.
+        :param obs: Observation
+        :return: The estimated Quantile-Value for each action.
+        """
+        quantile_values = self.quantile_net(self.extract_features(obs))
+        return quantile_values.view(-1, self.n_quantiles, self.action_space.n)
+
+    def _predict(self, observation: th.Tensor, deterministic: bool = True) -> th.Tensor:
+        q_values = self.forward(observation).mean(dim=1)
+        # Greedy action
+        action = q_values.argmax(dim=1).reshape(-1)
+        return action
+
+    def _get_data(self) -> Dict[str, Any]:
+        data = super()._get_data()
+
+        data.update(
+            dict(
+                net_arch=self.net_arch,
+                features_dim=self.features_dim,
+                n_quantiles=self.n_quantiles,
+                activation_fn=self.activation_fn,
+                features_extractor=self.features_extractor,
+                epsilon=self.epsilon,
+            )
+        )
+        return data
+
+
+class QRDQNPolicy(BasePolicy):
+    """
+    Policy class with Q-Value Net and target net for DQN
+    :param observation_space: Observation space
+    :param action_space: Action space
+    :param lr_schedule: Learning rate schedule (could be constant)
+    :param n_quantiles: Number of quantiles
+    :param net_arch: The specification of the policy and value networks.
+    :param activation_fn: Activation function
+    :param features_extractor_class: Features extractor to use.
+    :param features_extractor_kwargs: Keyword arguments
+        to pass to the features extractor.
+    :param normalize_images: Whether to normalize images or not,
+         dividing by 255.0 (True by default)
+    :param optimizer_class: The optimizer to use,
+        ``th.optim.Adam`` by default
+    :param optimizer_kwargs: Additional keyword arguments,
+        excluding the learning rate, to pass to the optimizer
+    """
+
+    def __init__(
+        self,
+        observation_space: gym.spaces.Space,
+        action_space: gym.spaces.Space,
+        lr_schedule: Schedule,
+        n_quantiles: int = 200,
+        net_arch: Optional[List[int]] = None,
+        activation_fn: Type[nn.Module] = nn.ReLU,
+        features_extractor_class: Type[BaseFeaturesExtractor] = FlattenExtractor,
+        features_extractor_kwargs: Optional[Dict[str, Any]] = None,
+        normalize_images: bool = True,
+        optimizer_class: Type[th.optim.Optimizer] = th.optim.Adam,
+        optimizer_kwargs: Dict[str, Any] = {"eps": 0.01 / 32},
```

Comment:
Please avoid mutable as default argument
why `0.01 / 32`, it depends on the minibatch size?

See https://github.com/satwikkansal/wtfpython#-beware-of-default-mutable-arguments

Reply:
>why 0.01 / 32, it depends on the minibatch size?

It's proposed in the paper (they swept it in  (0.01/32, 0.005/32, 0.001/32)), and 32 is the minibatch size.

Reply:
ok, probably put a comment too then ;)

Reply:
What is the best way to implement default "eps"?

Reply:
>What is the best way to implement default "eps"?


there are different ways...
we do it like that in A2C: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/a2c/a2c.py#L111

Reply:
Thank you :)